### ISSUE TITLE
Add support in BroadQuery to narrow the match to just phone or email 

### ISSUE THREAD

**Issue by vestrel00** [vestrel00/contacts-android#197]
@wo-ist-henry in #196 made me realize that the `BroadQuery` API is missing an important feature!

The `BroadQuery` API lets the Contacts Provider use its special matching algorithms to match contacts with a given search string. Contact data matched could be anything; **name, email, phone, address, organization, note, etc**...

## Problem

Currently, the `BroadQuery` API does not allow consumers to narrow the search scope to just phones or emails, which are probably very common use cases when performing searches. 

Sure, consumers can use `Query` to perform more custom, advanced matching. However, the results may differ from the custom matching algorithm of the Contacts Provider. Getting the `Query` API to return the same results as the matching algorithm of the Contacts Provider may also be impossible because not all phone numbers have a non-null `normalizedNumber` as pointed out in #196.

## Solution

Add support for narrowing the search scope in `BroadQuery` to just phones or emails. 

The additional API could look like;

```kotlin
BroadQuery {
    fun match(match: Match): BroadQuery

    enum class Match {
        ANY, // Default
        PHONE,
        EMAIL
    }
}
```

**Breaking changes** will only be to rename `whereAnyContactDataPartiallyMatches` to `wherePartiallyMatches`. Removing "AnyContactData" from the function name is important to avoid confusion.

## Developer notes

Matching only by phone number or email address is possible thanks to the following filter Uris defined in `ContactsContract`, which exist for this specific purpose.

```kotlin
ContactsContract {
    Contacts { CONTENT_FILTER_URI } // Default used by BroadQuery
    CommonDataKinds {
        Phone { CONTENT_FILTER_URI }
        Email { CONTENT_FILTER_URI }
    }
} 
```

**These special filter URIs are only available for the phone and email common data kinds.**

Note that the `EMAIL` and `PHONE` additionally matches the contact display name. Official documentation;

![Screen Shot 2022-03-08 at 9 38 24 AM](https://user-images.githubusercontent.com/4332753/157312363-a56928cf-3f28-4fc2-a531-53e57e962197.png)

## Comparison table

I've done some preliminary testing on the differences between the different matching/filter algorithms. So, given the following contacts...

1. Display name: Robert Parr 
    - Email: bob@incredibles.com
2. Display name: Bob Parr
    - Email: incredible@android.com
3. Display name: Bob Dole
    - Phone: (718) 737-1991
4. Display name: vestrel00@gmail.com
    - Email: vestrel00@gmail.com
5. Display name: 646-123-4567
    - Phone: 646-123-4567
6. Display name: Secret agent.    
    - Address: Dole street
    - Company: 718
    - Note: Agent code is 646000. His skills are incredible!

Here are some search terms followed by matching contacts based on the type of `Match` used.

| **Search term**    | **ANY** | **PHONE** | **EMAIL** |
|--------------------|---------|-----------|-----------|
| bob                | 1, 2, 3 | 3         | 1, 2      |
| incredible         | 1, 2, 6 |           | 2         |
| android            | 2       |           |           |
| gmail              | 4       |           |           |
| .com               | 1, 2, 4 |           |           |
| @                  |         |           |           |
| 7187371991         | 3       | 3         |           |
| 7.1-8.7-3.7-19(91) | 3       | 3         |           |
| 646                | 5, 6    | 5         |           |
| 646-646            | 6       |           |           |
| 718                | 3, 6    | 3         |           |
| 1991               |         |           |           |
| 4567               |         |           |           |
| 000                |         |           |           |
| dole               | 3, 6    | 3         |           |

The above table gives us some insight on how sophisticated the matching (or search / indexing) algorithm is.

For the search term "bob",

- PHONE matches contact 3.
    - Display name matches and contact has a phone even though it does match.
- EMAIL matches contact 1, 2.
    - 1 has a matching email "bob". 2 is also matched because the name matches even though the email does not. On the other hand, 3 is NOT matched even though the name matches because 3 has no email. Adding an email to 3 will cause 3 to be matched.

For the search term "incredible",

- EMAIL matches 2 (incredible@android.com) but NOT 1 (bob@incredibles.com). 
    - This means that email matching does not use `contains` but rather a form of `startsWith`.

**TLDR**

- ANY matches any contact data; name, email, phone, address, organization, note, etc.
- EMAIL matches emails or (display name + any email)
- PHONE matches phones or (display name + any phone)
- EMAIL and PHONE matching is NOT as simple as using the `Query` API with `.where { [Email|Phone] contains searchTerm }`

**Comment by vestrel00** [comment]
This is released in https://github.com/vestrel00/contacts-android/releases/tag/0.2.0
